packages/contracts-bedrock/src/L1/DataAvailabilityChallenge.sol

```diff
@@ -236,6 +264,51 @@ contract DataAvailabilityChallenge is OwnableUpgradeable, ISemver {
 
         // emit an event to notify that the challenge status is now resolved
         emit ChallengeStatusChanged(challengedHash, challengedBlockNumber, ChallengeStatus.Resolved);
+
+        // distribute the bond among challenger, resolver and address(0)
+        _distributeBond(activeChallenge, preImage.length, msg.sender);
+    }
+
+    /// @notice Distribute the bond of a resolved challenge among the resolver, challenger and address(0).
+    ///         The challenger is refunded the bond amount exceeding the resolution cost.
+    ///         The resolver is refunded a percentage of the resolution cost based on the `resolverRefundPercentage` state variable.
+    ///         The remaining bond is burned by sending it to address(0).
+    /// @dev The resolution cost is approximated based on a fixed cost and variable cost depending on the size of the pre-image.
+    ///      The real resolution cost might vary, because calldata is priced differently for zero and non-zero bytes.
+    ///      Computing the exact cost adds too much gas overhead to be worth the tradeoff.
+    /// @param resolvedChallenge The resolved challenge in storage.
+    /// @param preImageLength The size of the pre-image used to resolve the challenge.
+    /// @param resolver The address of the resolver.
+    function _distributeBond(Challenge storage resolvedChallenge, uint256 preImageLength, address resolver) internal {
+        uint256 lockedBond = resolvedChallenge.lockedBond;
+        address challenger = resolvedChallenge.challenger;
+
+        // approximate the cost of resolving a challenge with the provided pre-image size
+        uint256 resolutionCost = (fixedResolutionCost + preImageLength * variableResolutionCost) * tx.gasprice;
+
+        // refund bond exceeding the resolution cost to the challenger
+        if(lockedBond > resolutionCost) {
+            balances[challenger] += lockedBond - resolutionCost;
+            lockedBond = resolutionCost;
+            emit BalanceChanged(challenger, balances[challenger]);
+        }
+
+        // refund a percentage of the resolution cost to the resolver (but not more than the locked bond)
+        uint256 resolverRefund = resolutionCost * resolverRefundPercentage / 100;
+        if(resolverRefund > lockedBond) {
+            resolverRefund = lockedBond;
+        }
+        if(resolverRefund > 0) {
+            balances[resolver] += resolverRefund;
+            lockedBond -= resolverRefund;
+            emit BalanceChanged(resolver, balances[resolver]);
+        }
+
+        // burn the remaining bond
+        if(lockedBond > 0) {
+            payable(address(0)).transfer(lockedBond);
+        }
+        resolvedChallenge.lockedBond = 0;
```

Comment:
My [reentrancy](https://solidity-by-example.org/hacks/re-entrancy/) brain panicked that `lockedBond` was set after transferring the ETH, but of course the zero address can't reentry so the order is not _super_ important.